src/Orc.Controls/Orc.Controls.Shared/Controls/DateRangePicker/ViewModels/DateRangePickerViewModel.cs

```diff
@@ -0,0 +1,252 @@
+        {
+
+        }
+        #endregion
+
+        #region Properties
+        public ObservableCollection<DateRange> Ranges
+        {
+            get { return _ranges; }
+            set
+            {
+                if (_ranges == value)
+                {
+                    return;
+                }
+
+                _ranges = value;
+
+                RaisePropertyChanged(nameof(Ranges));
+            }
+        }
+
+        public DateRange SelectedRange
+        {
+            get { return _selectedRange; }
+            set
+            {
+                if (UpdateSelectedRange(_selectedRange, value))
+                {
+                    OnSelectedRangeChanged();
+                }
+            }
+        }
+
+        public DateTime DateStart
+        {
+            get { return _dateStart; }
+            set
+            {
+                if (UpdateDateStart(_dateStart, new DateTime(value.Year, value.Month, value.Day, value.Hour, value.Minute, value.Second)))
+                {
+                    OnDateStartChanged();
+                }
+            }
+        }
+
+        public DateTime DateEnd
+        {
+            get { return _dateEnd; }
+            set
+            {
+                if (UpdateDateEnd(_dateEnd, new DateTime(value.Year, value.Month, value.Day, value.Hour, value.Minute, value.Second)))
+                {
+                    OnDateEndChanged();
+                }
+            }
+        }
+
+        public TimeSpan Value
+        {
+            get { return _value; }
+            set
+            {
+                if (UpdateValue(_value, value))
+                {
+                    OnValueChanged();
+                }
+            }
+        }
+        #endregion
+
+        #region Methods
+        private void OnSelectedRangeChanged()
+        {
+            RemoveTemporaryRanges();
+
+            if (_selectedRange != null)
+            {
+                var currentDateStart = _dateStart;
+                var currentDateEnd = _dateEnd;
+                var currentValue = _value;
+                var newDateStart = _selectedRange.Start;
+                var newDateEnd = _selectedRange.End;
+                var newValue = newDateEnd.Subtract(newDateStart);
+
+                UpdateDateStart(currentDateStart, newDateStart);
+                UpdateDateEnd(currentDateEnd, newDateEnd);
+                UpdateValue(currentValue, newValue);
+            }
+        }
+
+        private void OnDateStartChanged()
+        {
+            RemoveTemporaryRanges();
+
+            var currentValue = _value;
+            var newValue = _dateEnd.Subtract(_dateStart);
+
+            UpdateValue(currentValue, newValue);
+
+            var currentSelectedDateRange = _selectedRange;
+            var newSelectedDateRange = AddTemporaryRange();
+
+            UpdateSelectedRange(currentSelectedDateRange, newSelectedDateRange);
+        }
+
+        private void OnDateEndChanged()
```

Comment:
Well.... yes and no, yes to avoid code repeating, no because each one method has own OnChangedMethod, it would be easier to understand and change something in the future. To be honest, in very starting version of this control those methods was different.... it became exact after last bug fixes.